src/main/scala/firrtl/transforms/Flatten.scala

```diff
@@ -0,0 +1,113 @@
+// See LICENSE for license details.
+
+package firrtl
+package passes
+
+import firrtl.ir._
+import firrtl.Mappers._
+import firrtl.annotations._
+import scala.collection.mutable
+
+// Tags an annotation to be consumed by this pass
+object FlattenAnnotation {
+  def apply(target: Named): Annotation = Annotation(target, classOf[Flatten], "")
+
+  def unapply(a: Annotation): Option[Named] = a match {
+    case Annotation(named, t, _) if t == classOf[Flatten] => Some(named)
+    case _ => None
+  }
+}
+
+//  
+// Takes all annotations for module instances and inline the entire hierarchy of modules down from the given ones.
+// This transformation is based on InlineInstances transformation. More concretely, it fully reuses run method and overriding execute method to collect the right set of modules to inline 
+// Note: Inlining a module means inlining all its children module instances   
+class Flatten extends InlineInstances {
+   private def collectAnns(circuit: Circuit, anns: Iterable[Annotation]): (Set[ModuleName], Set[ComponentName]) =
+     anns.foldLeft(Set.empty[ModuleName], Set.empty[ComponentName]) {
+       case ((modNames, instNames), ann) => ann match {
+         case FlattenAnnotation(CircuitName(c)) =>
+           (circuit.modules.collect {
+             case Module(_, name, _, _) if name != circuit.main => ModuleName(name, CircuitName(c))
+           }.toSet, instNames)
+         case FlattenAnnotation(ModuleName(mod, cir)) => (modNames + ModuleName(mod, cir), instNames)
+         case FlattenAnnotation(ComponentName(com, mod)) => (modNames, instNames + ComponentName(com, mod))
+         case _ => throw new PassException("Annotation must be InlineDeepAnnotation")
+       }
+     }
+
+   // Replicate the circuit such that we inline only the modules that are below hierarchy given by the annotations
+   // Along with replication we rewrite the original circuit to refer to the new modules that will be inlined
+   def duplicateSubCircuitsFromAnno(c : Circuit, mods:  Set[ModuleName], insts : Set[ComponentName], modsToInline : mutable.Set[ModuleName]) : Circuit = {
+     val modMap = c.modules.map(m => m.name->m) toMap
+     val seedMods = mutable.Map.empty[String, String]
+     val newModDefs= mutable.Set.empty[DefModule]
+     
+     // We start with rewriting original modules that contain annotations
+     def rewriteMod(parent : DefModule)(x : Statement) : Statement = x match {
+         case _: Block =>
+           x map rewriteMod(parent)
+         case WDefInstance(info, instName, moduleName, instTpe) =>
+           if (insts.contains(ComponentName(instName, ModuleName(parent.name, CircuitName(c.main))))
+            || mods.contains(ModuleName(parent.name, CircuitName(c.main)))) {
+             val newModName = moduleName+"_DEEPINLINE"
```

Comment:
Instead of this, you can use https://github.com/freechipsproject/firrtl/blob/master/src/main/scala/firrtl/Namespace.scala#L57 to create a module namespace.